certchain/certmanager/multicert_disk_cache.go

```diff
@@ -0,0 +1,181 @@
+// and reproduces an AugmentedChain. Read uses digest to retrieve the
+// AugmentedChain for that digest.  If digest is an empty string, Read returns
+// the latest revision of the AugmentedChain as referenced by LatestCertFilename
+// and LatestOCSPFilename.  Read only works only when d.CertDir is non-empty and
+// for cases where digest is an empty string, it only works when
+// d.LatestCertFilename and d.LatestOCSPFilename are non-empty,
+// otherwise returns an error. Read returns a multierror.Error
+// (hashicorp/go-multierror) to report as many problems as possible.
+func (d *MultiCertDiskCache) Read(digest string) (*certchain.AugmentedChain, error) {
+	var errs *multierror.Error
+
+	if d.CertDir == "" {
+		errs = multierror.Append(errs, errEmptyCertDir)
+	}
+
+	if err := errs.ErrorOrNil(); err != nil {
+		return nil, err
+	}
+
+	lock := flock.New(path.Join(d.CertDir, d.LockFilename))
+	errs = multierror.Append(errs, lock.RLock())
+
+	if err := errs.ErrorOrNil(); err != nil {
+		return nil, err
+	}
+
+	cert := path.Join(d.CertDir, digest+".pem")
+	ocsp := path.Join(d.CertDir, digest+".ocsp")
+	if digest == "" {
+		if d.LatestCertFilename == "" {
+			errs = multierror.Append(errs, errEmptyCertFilename)
+		}
+		if d.LatestOCSPFilename == "" {
+			errs = multierror.Append(errs, errEmptyOCSPFilename)
+		}
+		if err := errs.ErrorOrNil(); err != nil {
+			return nil, err
+		}
+		cert = path.Join(d.CertDir, d.LatestCertFilename)
+		ocsp = path.Join(d.CertDir, d.LatestOCSPFilename)
+	}
+	rawChain, err := certchainutil.ReadRawChainFile(cert)
+	errs = multierror.Append(errs, err)
+	ocspResp, err := certchainutil.ReadOCSPRespFile(ocsp)
+	errs = multierror.Append(errs, err)
+
+	var augChain *certchain.AugmentedChain
+	if errs.ErrorOrNil() == nil {
+		augChain = certchain.NewAugmentedChain(rawChain, ocspResp, nil)
+	}
+
+	errs = multierror.Append(errs, lock.Unlock())
+	return augChain, errs.ErrorOrNil()
+}
+
+// Write writes the certificate chain and the OCSP response from ac into local
+// files. It returns a multierror.Error (hashicorp/go-multierror) to report as
+// many problems as possible.
+func (d *MultiCertDiskCache) Write(ac *certchain.AugmentedChain) error {
+	if d.CertDir == "" {
+		return nil
+	}
+
+	var errs *multierror.Error
+
+	lock := flock.New(path.Join(d.CertDir, d.LockFilename))
+	errs = multierror.Append(errs, lock.Lock())
+
+	if err := errs.ErrorOrNil(); err != nil {
+		return err
+	}
+
+	errs = multierror.Append(errs, d.writeCert(ac))
+	errs = multierror.Append(errs, d.writeOCSP(ac))
+	errs = multierror.Append(errs, lock.Unlock())
+
+	return errs.ErrorOrNil()
+}
+
+func (d *MultiCertDiskCache) writeCert(ac *certchain.AugmentedChain) error {
+	if d.CertDir == "" || d.LatestCertFilename == "" {
+		return nil
+	}
+
+	buf := new(bytes.Buffer)
+	if err := ac.WritePEM(buf); err != nil {
+		return err
+	}
+
+	certPath := path.Join(d.CertDir, ac.Digest+".pem")
+	if err := renameio.WriteFile(certPath, buf.Bytes(), 0600); err != nil {
```

Comment:
Optional: It's probably not a big deal, but I'm not sure if certificates and OCSP responses are as sensitive as 0600.

Reply:
Ack.